### DISCUSSION TITLE
True way to create http server

### DISCUSSION THREAD

**Discussion by iAxel** [swoole/swoole-src#4719] [Q&A]
Hello everyone, please tell me how to create an http server correctly.

What do I need to know to create a http server?

What is the best mode to run the http server in?

What events are available for the http server?

What settings are required for the http server to work correctly?

What is the difference between SWOOLE_BASE and SWOOLE_PROCESS?

Please share your experience, thanks in advance :)

I know that there is documentation, but because of the language barrier, it's hard for me to understand it.

**Comment by iAxel** [comment]
Now I am creating a http server in a similar way in SWOOLE_BASE mode, it is important for me to understand if I am implementing this code correctly?
```php
<?php

declare(strict_types=1);

use Swoole\Http\Server;
use Swoole\Http\Request;
use Swoole\Http\Response;

$server = new Server('0.0.0.0', 9000, SWOOLE_BASE, SWOOLE_SOCK_TCP);

$server->set([
    'http_parse_post' => true,
    'http_parse_files' => false,
    'http_parse_cookie' => false,
    'http_compression' => true,
    'http_compression_level' => 1,
    'compression_min_length' => 128,
]);

$server->on('start', function (Server $server) {
    echo "PHP Http server running at http://$server->host:$server->port" . PHP_EOL;
});

$server->on('request', function (Request $request, Response $response) {
    $response->header('Content-Type', 'application/json');

    $response->status(200);

    $response->end(
        json_encode([
            'message' => 'Hello, world!',
        ]),
    );
});

$server->start();
```

**Reply by fakharak** [reply]
- As compared to SWOOLE_BASE , SWOOLE_PROCESS mode has two additional processes which are Master Process and Manager Process. Master Process contains multiple Event-loops for the service you develop whereas Manager Process manages the creation / removal of (I/O-intensive) Worker Processes. (I/O-intensive) Worker Processes is the process in which your Co-routines are run. 
Note: Swoole allows you to create Task Worker Process and hook callback in them to run CPU-intensie tasks, asynchronously. This mode is suitable for usecases which require more reliability like for larger and long-running service. This is default mode (when no mode is passed as argument)

- SWOOLE_BASE has one single process with a single Event Loop which also contains your Coroutine Context to run your (Asynchronous code snippets, called as) "Coroutines". This model is same as that of NodeJS. This model is faster than SWOOLE_PROCESS so it is suitable for faster real-time response but as you know if this single Process fails your whole service will crash and will need to be reloaded. So this mode is suitable for small jobs like in IoT / real-time Singal-processing. Single-process mode is also fitting to Containerization as each linux-container / docker-container also hosts only one single process. (My persoanl preference is multi-processing / parallel processing model of Swoole over "single-process mode for Containerisation" for several good reasons). 

In both modes, you can still create separate Task Workers and communicate with them asynchronously. A 'Tasks Worker' can be utilised for running CPU-intensive / long-running task.

**Reply by iAxel** [reply]
Thank you for answer!
